cpp/src/strings/convert/convert_datetime.cu

```diff
@@ -523,54 +542,53 @@ struct datetime_formatter {
     timeparts[TP_DAY]   = day;
     if (units == timestamp_units::days) return;
 
-    // done with date
-    // now work on time
-    int64_t hour = timestamp, minute = timestamp, second = timestamp;
+    // done with date, now work on time
+
     if (units == timestamp_units::hours) {
-      timeparts[TP_HOUR] = static_cast<int32_t>(hour % 24);
+      timeparts[TP_HOUR] = modulo_time(timestamp, 24);
       return;
     }
-    hour = hour / 60;
     if (units == timestamp_units::minutes) {
-      timeparts[TP_HOUR]   = static_cast<int32_t>(hour % 24);
-      timeparts[TP_MINUTE] = static_cast<int32_t>(minute % 60);
+      auto const hour      = scale_time(timestamp, 60);
+      timeparts[TP_HOUR]   = modulo_time(hour, 24);
+      timeparts[TP_MINUTE] = modulo_time(timestamp, 60);
       return;
     }
-    hour   = hour / 60;
-    minute = minute / 60;
     if (units == timestamp_units::seconds) {
-      timeparts[TP_HOUR]   = static_cast<int32_t>(hour % 24);
-      timeparts[TP_MINUTE] = static_cast<int32_t>(minute % 60);
-      timeparts[TP_SECOND] = static_cast<int32_t>(second % 60);
+      auto const hour      = scale_time(timestamp, 3600);
+      auto const minute    = scale_time(timestamp, 60);
+      timeparts[TP_HOUR]   = modulo_time(hour, 24);
+      timeparts[TP_MINUTE] = modulo_time(minute, 60);
+      timeparts[TP_SECOND] = modulo_time(timestamp, 60);
       return;
     }
-    hour   = hour / 1000;
-    minute = minute / 1000;
-    second = second / 1000;
+
+    // common utility for setting time from a subsecond unit value
+    auto subsecond_fn = [&](int64_t hour, int64_t minute, int64_t second, int64_t subsecond_base) {
+      timeparts[TP_HOUR]      = modulo_time(scale_time(hour, 3600), 24);
+      timeparts[TP_MINUTE]    = modulo_time(scale_time(minute, 60), 60);
+      timeparts[TP_SECOND]    = modulo_time(second, 60);
+      timeparts[TP_SUBSECOND] = modulo_time(timestamp, subsecond_base);
+    };
+
+    auto hour   = timestamp / 1000;
+    auto minute = timestamp / 1000;
+    auto second = timestamp / 1000;
     if (units == timestamp_units::ms) {
-      timeparts[TP_HOUR]      = static_cast<int32_t>(hour % 24);
-      timeparts[TP_MINUTE]    = static_cast<int32_t>(minute % 60);
-      timeparts[TP_SECOND]    = static_cast<int32_t>(second % 60);
-      timeparts[TP_SUBSECOND] = static_cast<int32_t>(timestamp % 1000);
+      subsecond_fn(hour, minute, second, 1000);
       return;
     }
     hour   = hour / 1000;
```

Comment:
The change made for timestamp_units::seconds and before which does away with cascading division of hours and minutes is a welcome improvement. Why not do the same here.